pkg/controllers/machine/lifecycle/registration_timeout.go

```diff
@@ -0,0 +1,60 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lifecycle
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/metrics"
+)
+
+type RegistrationTimeout struct {
+	clock      clock.Clock
+	kubeClient client.Client
+}
+
+// registrationTimeout is a heuristic time that we expect the node to register within
+// If we don't see the node within this time, then we should delete the machine and try again
+const registrationTimeout = time.Minute * 15
+
+func (r *RegistrationTimeout) Reconcile(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	if machine.StatusConditions().GetCondition(v1alpha5.MachineRegistered).IsTrue() {
+		return reconcile.Result{}, nil
+	}
+	if !machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).IsTrue() {
+		return reconcile.Result{}, nil
+	}
+	if r.clock.Since(machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).LastTransitionTime.Inner.Time) < registrationTimeout {
+		return reconcile.Result{RequeueAfter: registrationTimeout - r.clock.Since(machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).LastTransitionTime.Inner.Time)}, nil
+	}
+	// Delete the machine if we believe the machine won't register since we haven't seen the node
+	if err := r.kubeClient.Delete(ctx, machine); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	logging.FromContext(ctx).With("timeout", registrationTimeout).Debugf("terminating machine since node hasn't registered within registration timeout")
```

Comment:
Should this log come before the action? Otherwise it should be `terminated`

Reply:
I was thinking "terminating" since the finalizer would still have to clean the thing up. I'm fine either way though if you think that the gerund formatting is wrong here

Reply:
Do we have a common practice for whether or not we log before or after taking the action?